src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -29,39 +29,34 @@ internal sealed partial class AssetProvider(Checksum solutionChecksum, SolutionA
     private readonly SolutionAssetCache _assetCache = assetCache;
     private readonly IAssetSource _assetSource = assetSource;
 
-    private T GetRequiredAsset<T>(Checksum checksum)
-    {
-        Contract.ThrowIfTrue(checksum == Checksum.Null);
-        Contract.ThrowIfFalse(_assetCache.TryGetAsset<T>(checksum, out var asset));
-        return asset;
-    }
-
     public override async ValueTask<T> GetAssetAsync<T>(
         AssetHint assetHint, Checksum checksum, CancellationToken cancellationToken)
     {
         Contract.ThrowIfTrue(checksum == Checksum.Null);
         if (_assetCache.TryGetAsset<T>(checksum, out var asset))
             return asset;
 
-        using var _ = PooledHashSet<Checksum>.GetInstance(out var checksums);
+        using var _1 = PooledHashSet<Checksum>.GetInstance(out var checksums);
         checksums.Add(checksum);
 
-        await this.SynchronizeAssetsAsync(assetHint, checksums, cancellationToken).ConfigureAwait(false);
+        using var _2 = PooledDictionary<Checksum, object>.GetInstance(out var results);
+        await this.SynchronizeAssetsAsync(assetHint, checksums, results, cancellationToken).ConfigureAwait(false);
 
-        return GetRequiredAsset<T>(checksum);
+        return (T)results[checksum];
     }
 
-    public async ValueTask<ImmutableArray<ValueTuple<Checksum, T>>> GetAssetsAsync<T>(
+    public async ValueTask<ImmutableArray<(Checksum checksum, T asset)>> GetAssetsAsync<T>(
         AssetHint assetHint, HashSet<Checksum> checksums, CancellationToken cancellationToken)
     {
+        using var _1 = PooledDictionary<Checksum, object>.GetInstance(out var results);
+
         // bulk synchronize checksums first
         var syncer = new ChecksumSynchronizer(this);
-        await syncer.SynchronizeAssetsAsync(assetHint, checksums, cancellationToken).ConfigureAwait(false);
+        await syncer.SynchronizeAssetsAsync(assetHint, checksums, results, cancellationToken).ConfigureAwait(false);
 
-        // this will be fast since we actually synchronized the checksums above.
-        using var _ = ArrayBuilder<ValueTuple<Checksum, T>>.GetInstance(checksums.Count, out var list);
-        foreach (var checksum in checksums)
-            list.Add(ValueTuple.Create(checksum, GetRequiredAsset<T>(checksum)));
+        using var _2 = ArrayBuilder<ValueTuple<Checksum, T>>.GetInstance(checksums.Count, out var list);
```

Comment:
>using var _2 = ArrayBuilder<ValueTuple<Checksum, T>>.GetInstance(checksums.Count, out var list); [](http://example.com/codeflow?start=8&length=96)

Thoughts on just newing up an array and using ImmutableCollectionsMarshal to get it into an ImmutableArray for return?

Reply:
done.